Sources/SignalRClient/HubConnection.swift

```diff
@@ -339,6 +406,92 @@ public actor HubConnection {
             self.callback = callback
         }
     }
+
+    private struct DefaultInvocationBinder : InvocationBinder, @unchecked Sendable {
+        private let lock = DispatchSemaphore(value: 1)
+        private var subscriptionHandlers: [String: InvocationEntity] = [:]
+        private var returnValueHandler: [String: Any.Type] = [:]
+
+        mutating func registerSubscription(methodName: String, types: [Any.Type], handler: @escaping ([Any]) async throws -> Void) {
+            lock.wait()
+            defer {lock.signal()}
+            subscriptionHandlers[methodName] = InvocationEntity(types: types, callback: handler)
+        }
+
+        mutating func removeSubscrioption(methodName: String) {
+            lock.wait()
+            defer {lock.signal()}
+            subscriptionHandlers[methodName] = nil
+        }
+
+        mutating func registerInvocation(invocationId: String, types: Any.Type) {
+            lock.wait()
+            defer {lock.signal()}
+            returnValueHandler[invocationId] = types
+        }
+
+        mutating func removeInvocation(invocationId: String) {
+            lock.wait()
+            defer {lock.signal()}
+            returnValueHandler[invocationId] = nil
+        }
+
+        func getHandler(methodName: String) -> (([Any]) async throws -> Void)? {
+            lock.wait()
+            defer {lock.signal()}
+            return subscriptionHandlers[methodName]?.callback
+        }
+
+        func getReturnType(invocationId: String) -> (any Any.Type)? {
+            lock.wait()
+            defer {lock.signal()}
+            return returnValueHandler[invocationId]
+        }
+
+        func getParameterTypes(methodName: String) -> [any Any.Type] {
+            lock.wait()
+            defer {lock.signal()}
+            return subscriptionHandlers[methodName]?.types ?? []
+        }
+
+        func getStreamItemType(streamId: String) -> (any Any.Type)? {
+            lock.wait()
+            defer {lock.signal()}
+            return nil // not implemented
+        }   
+    }
+
+    private actor InvocationHandler {
+        private let lock = DispatchSemaphore(value: 1)
+        private var invocations: [String: TaskCompletionSource<CompletionMessage>] = [:]
+        private var id = 0
+
+        func create() async -> (String, TaskCompletionSource<CompletionMessage>) {
+            let newId = id + 1
+            id = newId
+            let tcs = TaskCompletionSource<CompletionMessage>()
+            invocations[String(newId)] = tcs
+            return (String(newId), tcs)
+        }
+
+        func setResult(message: CompletionMessage) async {
+            if let tcs = invocations[message.invocationId!] {
```

Comment:
Looks the `invocationId` of `CompletionMessage` should not be optional

Reply:
Yes, just because they inherit from HubInvocationMessage whose invocationId is nullable